File: Backtracking/39_Combination_Sum/solution39.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> currentCombination;
        
        backtrack(candidates, target, 0, currentCombination, result);
        
        return result;
    }
    
private:
    void backtrack(vector<int>& candidates, int target, int startIndex, vector<int>& currentCombination, vector<vector<int>>& result) {
        if (target == 0) {
            result.push_back(currentCombination);
            return;
        }
        
        for (int i = startIndex; i < candidates.size(); ++i) {
            if (candidates[i] > target) continue;
            currentCombination.push_back(candidates[i]);
            
            backtrack(candidates, target - candidates[i], i, currentCombination, result);
            
            currentCombination.pop_back();
        }
    }
};

// Runtime1: 0ms, beats 100%
// Memory1: 12.62MB, beats 57.93%

// Runtime2: 3ms, beats 79.53%
// Memory2: 12.54MB, beats 81.60%

// Runtime3: 0ms, beats 100%
// Memory3: 12.60MB, beats 81.60%

File: Backtracking/39_Combination_Sum/solution39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        
        def backtrack(remaining, start, combination):
            if remaining == 0:
                result.append(combination[:])
                return
            
            for i in range(start, len(candidates)):
                if candidates[i] > remaining:
                    continue
                
                combination.append(candidates[i])
                backtrack(remaining - candidates[i], i, combination)
                combination.pop()
    
        backtrack(target, 0, [])    
        return result

# Runtime1: 39ms, beats 97.59%
# Memory1: 16.76MB, beats 11.69%

# Runtime2: 31ms, beats 99.91%
# Memory2: 16.71MB, beats 12.34%

# Runtime3: 39ms, beats 97.61%
# Memory3: 16.72MB, beats 12.34%